source/common/network/io_socket_error_impl.cc

```diff
@@ -0,0 +1,44 @@
+#include "common/network/io_socket_error_impl.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Network {
+
+Api::IoError::IoErrorCode IoSocketError::getErrorCode() const {
+  switch (errno_) {
+  case EAGAIN:
+    ASSERT(this == IoSocketError::getIoSocketEagainInstance(),
+           "Didn't use getIoSocketEagainInstance() to generate `Again`.");
+    return IoErrorCode::Again;
+  case ENOTSUP:
+    return IoErrorCode::NoSupport;
+  case EAFNOSUPPORT:
+    return IoErrorCode::AddressFamilyNoSupport;
+  case EINPROGRESS:
+    return IoErrorCode::InProgress;
+  case EPERM:
+    return IoErrorCode::Permission;
+  default:
+    return IoErrorCode::UnknownError;
+  }
+}
+
+std::string IoSocketError::getErrorDetails() const { return ::strerror(errno_); }
+
+// static.
+IoSocketError* IoSocketError::getIoSocketEagainInstance() {
+  static auto* instance = new IoSocketError(EAGAIN);
```

Comment:
nit: can use CONSTRUCT_ON_FIRST_USE here

Reply:
not exactly same with CONSTRUCT_ON_FIRST_USE macro. We want a non-const to be returned to construct a unique_ptr.

Reply:
You could just take the address of what is returned, which I think would be better, but it's not a big deal either way.

Reply:
do you mean casting from `const IoSocketError*` to `IoSocketError*`?

Reply:
I mean have it return `IoSocketError&` and then do `&getIoSocketEagainInstance()` but I see that we have a const difference, so I guess that won't work.